---
author: "Gunjan"
output: html_document
---
  ```{r}
# transfer all .bed files in same directory and then do : grep "" .bed > simBed.txt. Do similarly for all the files

# use the following on command line
# Rscript *.md \
# -d coverage \
# -t variant type
# -s simBedfile \
# -p pindel \
# -c cnvnator \
# -l lumpy \
# -b svaba

# example as follows:
# cp /home/ggg256/scripts/simulation_del_analysis.Rmd .
# Rscript -e "library(knitr); knit('simulation_del_analysis.Rmd')" -p ${coverage}_pindel.txt \
# -c ${coverage}_cnv.txt \
# -l ${coverage}_lumpy.txt -b ${coverage}_svaba.txt \
# -s ${path_to_simBed_file} -d ${coverage} -t ${variant_type}
```

```{r,echo=FALSE,warning=FALSE,results='hide',message=FALSE}
#source("https://bioconductor.org/biocLite.R")
#biocLite("GenomicRanges")
require(GenomicRanges)

#install.packages("tidyr",repos = "http://cran.us.r-project.org")
#install.packages("dplyr",repos = "http://cran.us.r-project.org")
library(tidyr)
library(dplyr)
library(optparse)
library(knitr)
```

```{r reading in the files,echo=FALSE}
# reading in files and clean-up

option_list <- list(
  make_option(c("-d", "--depth"),action = "store",default = NULL,help="depth"),
  make_option(c("-t", "--type"),action = "store",default = NULL,help="variant type"),
  make_option(c("-b", "--svaba"),action = "store",default = NULL,help="svaba file"),
  make_option(c("-p", "--pindel"),action = "store",default = NULL,help="pindel file"),
  make_option(c("-c", "--cnvnator"),action = "store",default = NULL,help="cnvnator file"),
  make_option(c("-l", "--lumpy"),action = "store",default = NULL,help="lumpy file"),
  make_option(c("-s", "--simBed"),action = "store",default = NULL,help="simBed file"))

opt = parse_args(OptionParser(option_list=option_list))

simBed <- read.csv(opt$simBed,sep="\t",comment.char="",header=FALSE,stringsAsFactors = FALSE)
simBed <- (separate(data = simBed, col =V1, into = c("sim", "chrom"), sep = ".bed:"))

pind <- read.csv(opt$p,sep=" ",comment.char="",header=FALSE,stringsAsFactors = FALSE)
pind <- (separate(data = pind , col =V1, into = c("sim", "chrom"), sep = "_pindel.txt:"))
colnames(pind) <- c("V1","V2","V3","V4","V5","V6","V7")
p_provide <- pind[(pind$V2=="Provide"),]
pindel <- (pind %>% anti_join(p_provide))

cnvnator <- read.csv(opt$c, sep=" ",comment.char="",header=FALSE,stringsAsFactors = FALSE)
cnvnator <-(separate(data = cnvnator , col =V1, into = c("sim", "chrom"), sep = "_cnv.txt:"))

lumpy <- read.csv(opt$l, sep=" ",comment.char="",header=FALSE,stringsAsFactors = FALSE)
lumpy <-(separate(data = lumpy , col =V1, into = c("sim", "chrom"), sep = "_lumpy.txt:"))

svaba <- read.csv(opt$svaba, sep=",",comment.char="",header=FALSE,stringsAsFactors = FALSE)
svaba <- svaba[complete.cases(svaba),]
svaba <-(separate(data = svaba , col =V1, into = c("sim", "chrom"), sep = "_svaba.txt:"))
svaba$V4 <- abs(svaba$V2)+abs(svaba$V3)

depth=opt$d
type=opt$t
```

```{r data structures,echo=FALSE,message=FALSE}
colnames(pindel)<-c("sim","svtype","chrom.1","start","end","size","supporting.reads")
colnames(cnvnator)<-c("sim","svtype","chrom.1","start","end","size","normRDsignal")
colnames(lumpy)<-c("sim","svtype","chrom.1","start","end","size")
colnames(simBed) <- c("sim","chrom.1","start","chrom.2","end","svtype")
colnames(svaba) <- c("sim","chrom.1","start","size","end")

simBed$sim <- as.character(simBed$sim)
simBed$chrom.1 <- as.character(simBed$chrom.1)
simBed$chrom.2 <- as.character(simBed$chrom.2)
simBed$svtype <- as.character(simBed$svtype)
simBed$start <- as.integer(simBed$start)
simBed$end <- as.integer(simBed$end)

pindel$sim <- as.character(pindel$sim)
pindel$svtype <- as.character(pindel$svtype)
pindel$chrom.1 <- as.character(pindel$chrom.1)
pindel$start <- as.integer(pindel$start)
pindel$end <- as.integer(pindel$end)
pindel$size <- as.integer(pindel$size)
pindel$supporting.reads <- as.integer(pindel$supporting.reads)

cnvnator$sim <- as.character(cnvnator$sim)
cnvnator$svtype <- as.character(cnvnator$svtype)
cnvnator$chrom.1 <- as.character(cnvnator$chrom.1)
cnvnator$start <- as.integer(cnvnator$start)
cnvnator$end <- as.integer(cnvnator$end)
cnvnator$size <- as.integer(cnvnator$size)
cnvnator$normRDsignal <- as.numeric(cnvnator$normRDsignal)

lumpy$sim <- as.character(lumpy$sim)
lumpy$svtype <- as.character(lumpy$svtype)
lumpy$chrom.1 <- as.character(lumpy$chrom.1)
lumpy$start <- as.integer(lumpy$start)
lumpy$end <- as.integer(lumpy$end)
lumpy$size <- as.integer(abs(lumpy$size))

svaba$sim <- as.character(svaba$sim)
svaba$chrom.1 <- as.character(svaba$chrom.1)
svaba$start <- as.integer(svaba$start)
svaba$size <- as.integer(svaba$size)
svaba$end <- as.integer(svaba$end)
```

```{r filtering,echo=FALSE,message=FALSE}
# selecting for "deletions"
pindel.del <- pindel[pindel$svtype=="D" & (pindel$size)>50 & (pindel$supporting.reads)>10 , ]
cnv.del <- cnvnator[cnvnator$svtype=="deletion" & (cnvnator$size)>50 & (cnvnator$normRDsignal) < 0.2 ,]
lumpy.del <- lumpy[lumpy$svtype=="DEL" & (abs(lumpy$size))>50 ,]
svaba.del <- svaba[abs(svaba$size)>50,]
sim.del <- simBed[simBed$svtype=="DEL",]
```


```{r Granges,warning=FALSE,echo=FALSE,message=FALSE}
# this function would convert input dataframe to a GRange object
gRange.object <- function (object){
  sim <- object$sim
  chr <- object$chrom.1 
  start <- object$start
  end <- object$end
  gR <- GRanges(seqnames = paste0(sim,chr), ranges = IRanges(start=start, end=end))
  return (gR)
}

# this function would estimate performance parameters for pindel/lumpy/CNVnator/Svaba using Grange objects

scores <- function (query,subject,algorithm,coverage) {
  sim <- gRange.object(subject)
  test <- gRange.object(query)
  res <- countOverlaps(test,sim,type="any")
  TP <- length(res[!res==0])
  FP <- length(res[res==0])
  present <- dim(subject)[1]
  detected <- dim(query)[1]
  FDR <- round(((length(res[res==0]))/(length(res))),2) 
  precision <- round(((TP)/(TP+FP)),2)
  
  res1 <- countOverlaps(sim,test,type="any")
  FN <- (length(res1[res1==0]))
  sensitivity = round(((TP)/(TP+FN)),2)
  
  Fscore<-2*((precision*sensitivity)/(precision+sensitivity))
  Fscore <- round(Fscore,3)
  
  analysis <-t(as.data.frame(as.matrix(c(type,coverage,algorithm,present,detected,TP,FP,FN,FDR,precision,sensitivity,Fscore))))
  
  colnames(analysis) <- c("type","coverage","algorithm","#present","#detected","TP","FP","FN","FDR","precision","sensitivity","Fscore")
  return (analysis)
}

cnv.del.analysis<-scores(cnv.del,sim.del,"cnvnator",depth)
lumpy.del.analysis<-scores(lumpy.del,sim.del,"lumpy",depth)
pindel.del.analysis<-scores(pindel.del,sim.del,"pindel",depth)
svaba.del.analysis<-scores(svaba.del,sim.del,"svaba",depth)

# rbind only if variables have data
if (dim(cnv.del.analysis)[2]!=0){final <- rbind(cnv.del.analysis)}
if (dim(pindel.del.analysis)[2]!=0){final <- rbind(final,pindel.del.analysis)}
if (dim(lumpy.del.analysis)[2]!=0){final <- rbind(final,lumpy.del.analysis)}   
if (dim(svaba.del.analysis)[2]!=0){final <- rbind(final,svaba.del.analysis)}

rownames(final)<-c()
final <- as.data.frame(final)
final
write.csv(final, file = paste0(depth,type,".xls"))
```

```{r,warning=FALSE,message=FALSE,echo=FALSE}
# tabulate if an algorithm/tools detected a CNV

# converting to GRange object (This step could be skipped)
pindel.GR <- gRange.object(pindel.del)
cnv.GR <- gRange.object(cnv.del)
lumpy.GR <- gRange.object(lumpy.del)
svaba.GR <- gRange.object(svaba.del)
sim.GR <- gRange.object(sim.del)

# capture result for each algorithm/tool
pindel.vote <- sapply(countOverlaps(sim.GR,pindel.GR, type = "any"), function(x){if(x>0){return(2)}else{return(1)}})
cnv.vote <- sapply(countOverlaps(sim.GR,cnv.GR, type = "any"), function(x){if(x>0){return(2)}else{return(1)}})
lumpy.vote <- sapply(countOverlaps(sim.GR,lumpy.GR, type = "any"), function(x){if(x>0){return(2)}else{return(1)}})
svaba.vote <- sapply(countOverlaps(sim.GR,svaba.GR, type = "any"), function(x){if(x>0){return(2)}else{return(1)}})

# initialize to summarize 
votes <- NULL
votes <- as.data.frame(cbind(type,sim.del$sim, sim.del$chrom.1,sim.del$start,sim.del$end,
                             as.numeric(sim.del$end)-as.numeric(sim.del$start)))
colnames(votes) <- c("type","sim","chrom","start","end","size")

votes$pindel <- 0
votes$cnvnator <- 0
votes$lumpy <- 0
votes$svaba <- 0

if (length(pindel.vote)>1){votes$pindel <- pindel.vote} else {votes$pindel <- 0}
if (length(cnv.vote)>1){votes$cnvnator <- cnv.vote} else {votes$cnvnator <- 0}
if (length(lumpy.vote)>1){votes$lumpy <- lumpy.vote} else {votes$lumpy  <- 0}
if (length(svaba.vote)>1){votes$svaba <- svaba.vote} else {votes$svaba <- 0}

votes$pindel <- as.numeric(votes$pindel)
votes$lumpy <- as.numeric(votes$lumpy)
votes$cnvnator <- as.numeric(votes$cnvnator)
votes$svaba <- as.numeric(votes$svaba)

votes$pindel[votes$pindel==1] <- 0
votes$cnvnator[votes$cnvnator==1] <- 0
votes$lumpy[votes$lumpy==1] <- 0
votes$svaba[votes$svaba==1] <- 0

votes$pindel[votes$pindel==2] <- 1
votes$cnvnator[votes$cnvnator==2] <- 1
votes$lumpy[votes$lumpy==2] <- 1
votes$svaba[votes$svaba==2] <- 1

votes$call <- (votes$pindel + votes$cnvnator + votes$lumpy + votes$svaba)
votes <- votes[(order(votes$call,decreasing=TRUE)),]
MAIN.DF.1 <- data.frame()
MAIN.DF.1 <- subset(votes,select=c(pindel,cnvnator,lumpy,svaba))
MAIN.DF.1$truth <- 1
write.csv(votes, file = paste0(depth,type,"_vote",".xls"))
```
